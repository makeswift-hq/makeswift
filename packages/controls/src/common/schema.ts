import { z } from 'zod'

import { ResourceSchema } from '../resources'

const primitives = z.union([
  z.undefined(),
  z.null(),
  z.boolean(),
  z.number(),
  z.string(),
])

// See https://github.com/colinhacks/zod?tab=readme-ov-file#recursive-types
type Data = z.infer<typeof primitives> | Data[] | { [key: string]: Data }

export const data: z.ZodType<Data> = z.any()

export const colorData = z.object({
  swatchId: z.string(),
  alpha: z.number(),
})

export const resolvedColorData = z.object({
  swatch: ResourceSchema.swatch,
  alpha: z.number(),
})

export const deviceId = z.string()

export const deviceOverride = <T extends z.ZodTypeAny>(value: T) =>
  z.object({
    deviceId,
    value,
  })

export const responsiveValue = <T extends z.ZodTypeAny>(value: T) =>
  z.array(deviceOverride(value))

export const elementData = z.object({
  type: z.string(),
  key: z.string(),
  props: z.record(data),
})

export const elementReference = z.object({
  type: z.literal('reference'),
  key: z.string(),
  value: z.string(),
})

export const element = z.union([elementData, elementReference])
