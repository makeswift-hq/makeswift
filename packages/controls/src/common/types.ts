export type Device = string

export type DeviceOverride<T> = { deviceId: Device; value: T }

export type ResponsiveValue<T> = DeviceOverride<T>[]

export type ColorData = { swatchId: string; alpha: number }

export type Data = undefined | null | boolean | number | string | Data[] | { [key: string]: Data }
