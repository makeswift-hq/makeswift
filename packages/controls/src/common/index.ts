export * from './data-type-key'
export * from './types'