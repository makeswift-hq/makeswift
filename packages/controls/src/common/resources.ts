export type Swatch = {
  id: string
  hue: number
  saturation: number
  lightness: number
}
