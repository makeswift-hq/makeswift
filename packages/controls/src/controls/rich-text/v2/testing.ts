import { z } from 'zod'

import { type SerializedRecord } from '../../../serialization'
import { type DataType } from '../../associated-types'
import { serialize } from '../../definition'
import {
  ControlInstance,
  DefaultControlInstance,
  type SendMessage,
} from '../../instance'

import { RichTextPluginControl } from './plugin'
import { RichTextDefinition } from './rich-text'

export const renderedNode = Symbol('rendered node')

export type RenderedNode = typeof renderedNode

type UserConfig = z.infer<typeof Definition.schema.userConfig>

class Definition extends RichTextDefinition<RenderedNode> {
  createInstance(sendMessage: SendMessage<any>): ControlInstance<any> {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }

  toText(_data: DataType<RichTextDefinition<string>>): string {
    return ''
  }

  get pluginControls(): RichTextPluginControl[] {
    return []
  }

  pluginControlAt(_index: number) {
    return undefined
  }
}

export { Definition as RichTextDefinition }

export const RichText = (config?: UserConfig) => new Definition(config ?? {})

RichText.Mode = Definition.Mode
