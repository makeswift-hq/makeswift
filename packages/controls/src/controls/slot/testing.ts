import { SlotDefinition } from './slot'

export const renderedNode = Symbol('rendered node')

export type RenderedNode = typeof renderedNode

class Definition extends SlotDefinition<RenderedNode> {}

export function Slot(): Definition {
  return new Definition()
}

export { Definition as SlotDefinition }
