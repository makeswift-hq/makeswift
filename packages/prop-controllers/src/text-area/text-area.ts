import { z } from 'zod'
import { ControlDataTypeKey, Options, Types } from '../prop-controllers'
import { match } from 'ts-pattern'

const textAreaPropControllerDataV0Schema = z.string()

export type TextAreaPropControllerDataV0 = z.infer<
  typeof textAreaPropControllerDataV0Schema
>

export const TextAreaPropControllerDataV1Type =
  'prop-controllers::text-area::v1'

const textAreaPropControllerDataV1Schema = z.object({
  [ControlDataTypeKey]: z.literal(TextAreaPropControllerDataV1Type),
  value: z.string(),
})

export type TextAreaPropControllerDataV1 = z.infer<
  typeof textAreaPropControllerDataV1Schema
>

export const textAreaPropControllerDataSchema = z.union([
  textAreaPropControllerDataV0Schema,
  textAreaPropControllerDataV1Schema,
])

export type TextAreaPropControllerData = z.infer<
  typeof textAreaPropControllerDataSchema
>

export type TextAreaOptions = Options<{
  preset?: TextAreaPropControllerData
  label?: string
  rows?: number
}>

type TextAreaDescriptorV0<_T = TextAreaPropControllerDataV0> = {
  type: typeof Types.TextArea
  options: TextAreaOptions
}

type TextAreaDescriptorV1<
  _T = TextAreaPropControllerData,
  U extends TextAreaOptions = TextAreaOptions,
> = {
  type: typeof Types.TextArea
  version: 1
  options: U
}

export type TextAreaDescriptor<_T = TextAreaPropControllerData> =
  | TextAreaDescriptorV0
  | TextAreaDescriptorV1

export type ResolveTextAreaPropControllerValue<T extends TextAreaDescriptor> =
  T extends TextAreaDescriptor ? string | undefined : never

/**
 * @deprecated Imports from @makeswift/prop-controllers are deprecated. Use
 * @makeswift/runtime/controls instead.
 */
export function TextArea(options: TextAreaOptions = {}): TextAreaDescriptorV1 {
  return { type: Types.TextArea, version: 1, options }
}

export function getTextAreaPropControllerDataString(
  data: TextAreaPropControllerData | undefined,
): string | undefined {
  return match(data)
    .with(
      { [ControlDataTypeKey]: TextAreaPropControllerDataV1Type },
      (v1) => v1.value,
    )
    .otherwise((v0) => v0)
}

export function createTextAreaPropControllerDataFromString(
  value: string,
  definition: TextAreaDescriptor,
): TextAreaPropControllerData {
  return match(definition)
    .with(
      { version: 1 },
      () =>
        ({
          [ControlDataTypeKey]: TextAreaPropControllerDataV1Type,
          value,
        } as const),
    )
    .otherwise(() => value)
}
