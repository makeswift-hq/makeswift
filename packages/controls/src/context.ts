import { type Data, type Element, type ElementData } from './common/types'

export type TranslationDto = Record<string, Data>
export type MergeTranslatableDataContext = {
  translatedData: TranslationDto
  mergeTranslatedData: (node: Element) => Element
}

export type ReplacementContext = {
  elementHtmlIds: Set<string>
  elementKeys: Map<string, string>
  swatchIds: Map<string, string>
  fileIds: Map<string, string>
  typographyIds: Map<string, string>
  tableIds: Map<string, string>
  tableColumnIds: Map<string, string>
  pageIds: Map<string, string>
  globalElementIds: Map<string, string>
  globalElementData: Map<string, ElementData>
}

export type SerializableReplacementContext = {
  elementHtmlIds?: string[]
  elementKeys?: Record<string, string>
  swatchIds?: Record<string, string>
  fileIds?: Record<string, string>
  typographyIds?: Record<string, string>
  tableIds?: Record<string, string>
  tableColumnIds?: Record<string, string>
  pageIds?: Record<string, string>
  globalElementIds?: Record<string, string>
  globalElementData?: Record<string, ElementData>
}

export function createReplacementContext(
  context: SerializableReplacementContext,
): ReplacementContext {
  const toMap = <V>(record?: Record<string, V>) =>
    new Map(Object.entries(record ?? {}))

  return {
    elementHtmlIds: new Set(context.elementHtmlIds),
    elementKeys: toMap(context.elementKeys),
    swatchIds: toMap(context.swatchIds),
    fileIds: toMap(context.fileIds),
    typographyIds: toMap(context.typographyIds),
    tableIds: toMap(context.tableIds),
    tableColumnIds: toMap(context.tableColumnIds),
    pageIds: toMap(context.pageIds),
    globalElementIds: toMap(context.globalElementIds),
    globalElementData: toMap(context.globalElementData),
  }
}

export type CopyContext = {
  replacementContext: ReplacementContext
  copyElement: (node: Element) => Element
}

export type MergeContext = {
  mergeElement(a: Element, b: Element): Element
}
