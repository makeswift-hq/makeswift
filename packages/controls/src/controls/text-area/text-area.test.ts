import { testDefinition } from '../../testing/test-definition'

import { ControlDataTypeKey } from '../../common'
import { MergeTranslatableDataContext, TranslationDto } from '../../context'

import { TextArea, TextAreaDefinition } from './text-area'

describe('TextArea', () => {
  describe('constructor', () => {
    test.each(['Toy Story', 'Cars', undefined])(
      'call with default value `%s` returns versioned definition',
      (value) => {
        expect(
          TextArea({
            label: 'TextArea',
            defaultValue: value,
          }),
        ).toMatchSnapshot()
      },
    )

    test('disallows extraneous properties', () => {
      TextArea({
        label: undefined,
        defaultValue: undefined,
        // @ts-expect-error
        extra: 'extra',
      })
    })
  })

  describe('assignability', () => {
    function assignTest(_def: TextAreaDefinition) {}
    assignTest(TextArea())
    assignTest(TextArea({ label: 'TextArea' }))
    assignTest(TextArea({ defaultValue: 'text' }))
    assignTest(TextArea({ defaultValue: 'text', rows: 3 }))
    assignTest(TextArea({ defaultValue: 'text' as string }))
    assignTest(TextArea({ defaultValue: undefined, rows: undefined }))
    assignTest(
      TextArea({ label: undefined, defaultValue: undefined, rows: undefined }),
    )
  })

  describe('getTranslatableData', () => {
    const def = TextArea({ label: 'TextArea' })
    test.each(['The Incredibles', def.toData('The Incredibles'), undefined])(
      'returns `%s` when data is `%s`',
      (data) => {
        const def = TextArea({ label: 'TextArea' })
        expect(def.getTranslatableData(data)).toBe(data)
      },
    )
  })

  describe('mergeTranslatedData', () => {
    const translationData: TranslationDto = {
      key: 'Le monde de nemo',
    }

    const mergeContext: MergeTranslatableDataContext = {
      translatedData: translationData,
      mergeTranslatedData: (el) => el,
    }

    const def = TextArea({ label: 'TextArea' })

    test('returns translated data when translated data is non-nullish', () => {
      const dataV0 = 'Finding Nemo'
      const dataV1 = {
        [ControlDataTypeKey]: 'text-area::v1' as const,
        value: 'Finding Nemo',
      }

      expect(
        def.mergeTranslatedData(dataV1, translationData, mergeContext),
      ).toBe(translationData)
      expect(
        def.mergeTranslatedData(dataV0, translationData, mergeContext),
      ).toBe(translationData)
    })

    test('returns data when translated data is nullish', () => {
      const dataV0 = 'Finding Nemo'
      const dataV1 = {
        [ControlDataTypeKey]: 'text-area::v1' as const,
        value: 'Finding Nemo',
      }

      expect(def.mergeTranslatedData(dataV1, undefined, mergeContext)).toBe(
        dataV1,
      )
      expect(def.mergeTranslatedData(dataV0, null, mergeContext)).toBe(dataV0)
    })
  })

  const invalidValues = [null, false, 5, []]

  testDefinition(
    TextArea({ defaultValue: 'Up', label: 'visible' }),
    ['Toy Story 3', "Monster's Inc."],
    invalidValues,
  )

  testDefinition(
    TextArea({ label: 'TextArea' }),
    ['WALL-E', 'Ratatouille', undefined],
    invalidValues,
  )
})
