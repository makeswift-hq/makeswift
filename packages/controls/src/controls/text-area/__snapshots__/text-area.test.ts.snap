// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea constructor call with default value \`Cars\` returns versioned definition 1`] = `
TextAreaDefinition {
  "config": {
    "defaultValue": "Cars",
    "label": "TextArea",
  },
  "version": 1,
}
`;

exports[`TextArea constructor call with default value \`Toy Story\` returns versioned definition 1`] = `
TextAreaDefinition {
  "config": {
    "defaultValue": "Toy Story",
    "label": "TextArea",
  },
  "version": 1,
}
`;

exports[`TextArea constructor call with default value \`undefined\` returns versioned definition 1`] = `
TextAreaDefinition {
  "config": {
    "defaultValue": undefined,
    "label": "TextArea",
  },
  "version": 1,
}
`;

exports[`TextArea definition w/ config {"defaultValue":"Up","label":"visible"} toData returns v1 data for \`Monster's Inc.\` when definition version is 1 1`] = `
{
  "@@makeswift/type": "text-area::v1",
  "value": "Monster's Inc.",
}
`;

exports[`TextArea definition w/ config {"defaultValue":"Up","label":"visible"} toData returns v1 data for \`Toy Story 3\` when definition version is 1 1`] = `
{
  "@@makeswift/type": "text-area::v1",
  "value": "Toy Story 3",
}
`;

exports[`TextArea definition w/ config {"label":"TextArea"} toData returns v1 data for \`Ratatouille\` when definition version is 1 1`] = `
{
  "@@makeswift/type": "text-area::v1",
  "value": "Ratatouille",
}
`;

exports[`TextArea definition w/ config {"label":"TextArea"} toData returns v1 data for \`WALL-E\` when definition version is 1 1`] = `
{
  "@@makeswift/type": "text-area::v1",
  "value": "WALL-E",
}
`;

exports[`TextArea definition w/ config {"label":"TextArea"} toData returns v1 data for \`undefined\` when definition version is 1 1`] = `undefined`;
